pvl/__init__.py

```diff
@@ -73,90 +34,164 @@
 ]
 
 
-def __create_decoder(cls, strict, **kwargs):
-    decoder = cls(**kwargs)
-    decoder.set_strict(strict)
-    return decoder
-
+def load(path, **kwargs):
+    """Returns a Python object from parsing the file at *path*.
 
-def load(stream, cls=PVLDecoder, strict=True, **kwargs):
-    """Deserialize ``stream`` as a pvl module.
+    :param path: an :class:`os.PathLike` which presumably has a
+        PVL Module in it to parse.
+    :param ``**kwargs``: the keyword arguments that will be passed
+        to :func:`loads()` and are described there.
 
-    :param stream: a ``.read()``-supporting file-like object containing a
-        module. If ``stream`` is a string it will be treated as a filename
+    If *path* is not an :class:`os.PathLike`, it will be assumed to be an
+    already-opened file object, and ``.read()`` will be applied
+    to extract the text.
 
-    :param cls: the decoder class used to deserialize the pvl module. You may
-        use the default ``PVLDecoder`` class or provide a custom sublcass.
-
-    :param **kwargs: the keyword arguments to pass to the decoder class.
+    If the :class:`os.PathLike` or file object contains some bytes
+    decodable as text, followed by some that is not (e.g. an ISIS
+    cube file), that's fine, this function will just extract the
+    decodable text.
     """
-    decoder = __create_decoder(cls, strict, **kwargs)
-    if isinstance(stream, six.string_types):
-        with open(stream, 'rb') as fp:
-            return decoder.decode(fp)
-    return decoder.decode(stream)
-
-
-def loads(data, cls=PVLDecoder, strict=True, **kwargs):
-    """Deserialize ``data`` as a pvl module.
-
-    :param data: a pvl module as a byte or unicode string
-
-    :param cls: the decoder class used to deserialize the pvl module. You may
-        use the default ``PVLDecoder`` class or provide a custom sublcass.
-
-    :param **kwargs: the keyword arguments to pass to the decoder class.
+    return loads(get_text_from(path), **kwargs)
+
+
+def get_text_from(path) -> str:
+    try:
```

Comment:
If you can figure out a way to un-nest it, that would be great.

Reply:
I think this would work, have not attempted to run it yet:
```python
def get_text_from(path) -> str:
    try:
        return Path(path).read_text()
    except UnicodeDecodeError:
        # This may be the result of an ISIS cube file (or anything else)
        # where the first set of bytes might be decodable, but once the
        # image data starts, they won't be, and the above tidy function
        # fails.
        pass  
    try:
        #So open the file as a bytestream, and read until
        # we can't decode.  We don't want to just run the .read_bytes()
        # method of Path, because this could be a giant file.
        with open(p, mode='rb') as f:
            return decode_by_char(f)
    except TypeError:
        # Not an os.PathLike, maybe it is an already-opened file object
        pass
    if path.readable():
        try:
            position = path.tell()
            s = path.read()
            if isinstance(s, bytes):
                path.seek(position)  # Reset after the previous .read():
                s = decode_by_char(path)
        except UnicodeDecodeError:
            # All of the bytes weren't decodeable, maybe the initial
            # sequence is (as above)?
            path.seek(position)  # Reset after the previous .read():
            s = decode_by_char(path)
    else:
        # Not a path, not an already-opened file.
        raise TypeError('Expected an os.PathLike or an already-opened '
                        'file object, but did not get either.')
    return s
```

Reply:
Thanks, but I don't think that will work, but this did give me an idea.  

It won't work because using `Path(path)` may emit a TypeError (if `path` isn't an os.PathLike), and the `.read_text()` might emit the UnicodeDecodeError if it is a file of mixed bytes.  Once I realized this, the answer is not nested trys (as you surmised), but a single try with two except handlers.

I also didn't like having the `seek()` and `decode_by_char()` logic doubled in the lower code block, so I also fixed that up, but that's *really* a bikeshed thing.

See what you think.